rfd/0180-sso-as-mfa-method.md

```diff
@@ -255,197 +403,322 @@ invariant, device trust must be enforced within the SSO MFA check.
+        Identity Provider ->> Teleport: send successful IdP auth response
+        Teleport ->> Teleport: save MFA token in SSO MFA session
+        Teleport ->> Identity Provider: return client callback URL<br/>w/ encode encrypted MFA token in query params
+        Identity Provider -> Client: redirect
+    end
+    Note over Client: Decode and decrypt MFA token<br/>from redirect query params
+    par mfa response validation
+        Client ->> Teleport: rpc GenerateUserSingleUseCerts<br/>w/ SSO MFA token
+        Teleport ->> Teleport: check MFA token against SSO MFA session
+        Teleport ->> Client: issue MFA verified certs
+    end
+    Client ->> Node: connect w/ MFA verified certs
+```
 
-```proto
-message MFAAuthenticateResponse {
-  oneof Response {
-    TOTPResponse TOTP = 2;
-    webauthn.CredentialAssertionResponse Webauthn = 3;
-    PrivilegeTokenResponse Token = 4;
-  }
+Caveat: this flow will be slightly different for the WebUI, where we use a CSRF
+token instead of a secret key, and save the MFA token as a web cookie to avoid
+redirecting the user in the browser.
+
+#### SSO MFA session
+
+When a user requests an SSO MFA challenge, an SSO MFA session data object will
+be created in the backend. Once the user completes the SSO MFA flow, the client
+and session data will gain a matching MFA token, which can be used by the client
+to verify itself as the owner of the session.
+
+Note: challenge extensions are included, so scopes and reuse are both available.
+See RFD 155 for details.
+
+```go
+// SSOMFASessionData SSO MFA Session data.
+type SSOMFASessionData struct {
+	// RequestID is the ID of the corresponding SSO Auth request, which is used to
+	// identity this session.
+	RequestID string `json:"request_id,omitempty"`
+	// Username is the Teleport username.
+	Username string `json:"username,omitempty"`
+	// MFAToken is a secret token set when the SSO flow is completed and returned to
+  // the user. The user can then use this token to prove they own the MFA session.
+	MFAToken string `json:"token,omitempty"`
+	// ConnectorID is id of the corresponding Auth connector.
+	ConnectorID string `json:"connector_id,omitempty"`
+	// ConnectorType is SSO type of the corresponding Auth connector (SAML, OIDC, or Github).
+	ConnectorType string `json:"connector_type,omitempty"`
+	// ChallengeExtensions are challenge extensions that apply to this SSO MFA session.
+	ChallengeExtensions *mfav1.ChallengeExtensions `json:"challenge_extensions,omitempty"`
 }
 ```
 
-Note: Privilege tokens used in this way must be consumed upon verification to
-ensure that exchanging a webauthn or totp token for a privilege token through
-`rpc CreatePrivilegeToken` has no significant security implications.
+#### MFA device
+
+Teleport will treat any user that originated from an mfa-enabled auth connector
+as if they have an SSO MFA device registered. This device is not stored in the
```

Comment:
I'm interested in how you want to implement this.

For example, if I call lib/services/local.IdentityService.GetMFADevices, do I see this device on the list?

Reply:
My concern there is that it's easy to forget about SSO devices, which could lead to bugs in places that take GetMFADevices to really mean "all devices" (which it does up until now!)

Reply:
Fair point, in that case we can include it in the `services/local` implementation, and also add a `GetUserSSOMFADevice` implementation there instead of in server. WDYT?

Reply:
So we'll both change GetMFADevices to include SSO and add a GetUserSSOMFADevice, is that it?

Something else to look out for are switches on the MFADevice.device oneof - I think they should be OK but it's worth checking.

Reply:
(Maybe add a few SSOMFADevices to tests here and there to be sure they behave well.)